### ISSUE TITLE
Metadata regarding type of ensemble

### ISSUE THREAD

**Issue by anders-kiaer** [equinor/fmu-dataio#299]
E.g. is it a
- history matching iteration ensemble (iter-0, ... iter-3...)
- prediction ensemble
- sensitivity ensemble

The "ensemble type" (better terminology might/probably exist) will be used by clients in order to filter out ensembles not relevant in a given analysis dashboard/pattern.

**Comment by anders-kiaer** [comment]
A bit more complex perhaps, but it would also be very useful to have metadata answering e.g. which history matching ensemble is a given prediction ensemble based on.

**Comment by anders-kiaer** [comment]
Some user stories on parent-child-linking of ensembles:
* You have a prediction run (which is a `RESTART` from an AHM run), and want to know the quality of the AHM-ensemble it was based on. With parent-child-linking metadata clients downstream can use this information to give the user this insight.
* You have a prediction run (which is a `RESTART` from an AHM run), and a user you want the time series and/or 3D grid to extend the whole time axis (history + prediction) and not just prediction.  With parent-child-linking metadata clients downstream can fetch data from both ensembles (history and prediction) and combine them in the presentation.
* You could have a "base prediction run" (e.g. going to 2030), and from there you have multiple different prediction scenarios reflecting different decisions to be made by the asset. If clients could e.g. query "which ensembles uses this ensemble as parent" that could be used to present a list of ensembles (future scenarios) based on the same base ensemble.
*[...]

**Comment by perolavsvendsen** [comment]
This links to #291 perhaps

**Comment by alifbe** [comment]
I have tried to run drogon_pred_ref.ert and found that the results for pred_ref ensemble are uploaded to iter-0. From discussion with @perolavsvendsen apparently, dataio assumed iter-0 as default ensemble name and used them to generate ID.

See case 01_drogon_ahm_sumo in sumo prod

**Comment by perolavsvendsen** [comment]
Yes, we currently don't get any information on type of ensembles within a case, and as far as I know no such definition exists either.

This obviously maps to ERT quite fast. If not, we would have to do something rule-based based on iteration names.

Is there a standard (convention) on iteration names?

**Comment by anders-kiaer** [comment]
No strict/official rules for iteration names to my knowledge (AHM runs typically go `iter-0`, `iter-1`, ..., `iter-x` and prediction cases often has `pred` in the name, but no guarantee `iter` and `pred` are used as substrings). 

I agree this information/metadata should come from ERT. Knowing the workflow method used by ERT for generating the ensemble would probably be a good start (https://ert.readthedocs.io/en/latest/reference/running_ert.html - `ensemble_smoother`, `es_mda`, `iterative_ensemble_smoother`, `ensemble_experiment`). Can/is this information exposed in some way @sondreso?

Within an assisted history matching run it would also be useful for clients of the data sets to know which ensembles are part of the same assisted history matching run (i.e. `iter-0` = prior, `iter-{max}` = posterior, and also the order of ensembles inbetween representing gradual updates from prior to posterior...).

Pinging @asnyv in case there are details / use cases you want to mention I haven't.

**Comment by asnyv** [comment]
Think @anders-kiaer has covered most of it, but for predictions I think we (at least me) often skip the term "pred" and use some more or less descriptive name dependent on whatever we are simulating. From a technical perspective: completely arbitrary.

Also: for a while I think it was fairly common to have a structure like:

History matching:
`some_ahm_case/realization-x/iter-y`

Predictions:
```
some_prediction_scenario/realization-x
some_other_prediction_scenario/realization-x
```

Now it seems like more are going towards a structure where the prediction case is placed on the "iteration level" like you mention, so:
```
some_ahm_case/realization-x/iter-y
some_ahm_case/realization-x/some_prediction_scenario
some_ahm_case/realization-x/some_other_prediction_scenario
```

The advantage with the latter is that it is clearer what the basis for the prediction is, whilst the advantage of the first one is that it is easier to see all cases in a folder structure + it is a more natural structure for models without any history. But many of the models without history have now ended with the structure: `some_prediction_scenario/realization-x/iter-0` as a more or less de facto standard, but I can't really say why 😅 My guess is that someone found it convenient as they could then reuse something they had hard-coded for AHM.

**Comment by perolavsvendsen** [comment]
Suggest we first try to accurately reflect the `name` in the outgoing metadata, not default to `iter-0` if it looks strange. The more challenging bit is probably the iteration `id`, which in turn maps to the iteration `uuid`.

I assume that ERT internally has an iteration ID which we cannot know if the iteration is called something other than e.g. `iter-0`. Every single instance of fmu-dataio will (currently) look at the file structure to determine what iteration we are in. The code has been structured in such a way that it should be possible to get this from ERT when it is available, but so far it is not.

An alternative is to take away all logic placed on the iteration `id` and only use the iteration `name`.

**Comment by perolavsvendsen** [comment]
Drafting a possible PR, quick and dirty 👆. This needs discussions, as I am a bit unsure of the consequences. But possibly, the iteration _name_ is a better option (outside the ERT context) than iteration _id_. E.g. the name is used as an identifier, not the ID. (This is similar to the current practice for the case.name, where no ID exists.

For the third similar object, `realization`, we are more dependent on the ID I guess.

**Comment by perolavsvendsen** [comment]
This is also a feature request from SSDL, ref discussions with @bous251

**Comment by anders-kiaer** [comment]
Thanks for the update @mferrera. Looks good - this metadata enables some of the use cases, and partially solves this issue. The remaining part is then relationships between ensembles. E.g.:

- Ensembles belonging to the same AHM run (`iter-0`, `iter-1`, `iter-2`, ..., `iter-N`).
- Ensembles which continue simulating ("restarts") from another ensemble (typical scenario is prediction runs which uses an AHM ensemble as starting point).

**Comment by perolavsvendsen** [comment]
> * Ensembles belonging to the same AHM run (`iter-0`, `iter-1`, `iter-2`, ..., `iter-N`).
This should be covered either via `fmu.case`, or probably better via `fmu.ert.experiment`.


> * Ensembles which continue simulating ("restarts") from another ensemble (typical scenario is prediction runs which uses an AHM ensemble as starting point).

This is much larger probably, and involves more than just metadata. Supporting restart should probably be a separate, larger issue/epic. Need to store the required files (not really scope for Sumo), amake them available on a disk area, etc. This feels like something which might be more in scope for ERT storage, but unsure.

**Comment by anders-kiaer** [comment]
Yes, the first bullet point is easiest to solve.

The second one has value also without additional files stored in Sumo. E.g.
- if you have a prediction ensemble in Sumo that was simulated from 2025 and onwards, the Sumo client could want to present the end user concatenated time series from that prediction ensemble together with the time series data in Sumo from the corresponding history matched ensemble it was restarted from.
- ...or you want to show the user which parameters where inherited from the AHM ensemble into the prediction ensemble (since it is a 1-1 correspondence between realizations in the two ensembles).

With that being said, the metadata to know which ensemble a given ensemble was restarted from is not owned by any software today (not even ERT), so perhaps we need the full solution like you point to @perolavsvendsen (being able to restart from previous run stored somewhere in a standard way) in order for `fmu-dataio` to even get hold of the required parent-child ensemble metadata.
